File: 2013 ACM-ICPC North American Qualifier/island.py
from collections import deque
import sys

class DSU:
    def __init__(self, n):
        self.parent = list(range(n + 1))
    
    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
    
    def unite(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            self.parent[root_i] = root_j

def main():
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    data = sys.stdin.read().splitlines()
    idx = 0
    map_num = 1
    first_map_output = True
    
    while idx < len(data):
        line = data[idx].strip()
        idx += 1
        
        # Skip empty lines
        while idx < len(data) and not line:
            line = data[idx].strip()
            idx += 1
        
        if not line:
            break
        
        if not first_map_output:
            print()
        first_map_output = False
        
        grid_lines = [line]
        while idx < len(data) and data[idx].strip():
            grid_lines.append(data[idx].strip())
            idx += 1
        
        R = len(grid_lines)
        C = len(grid_lines[0])
        
        grid = [list(line) for line in grid_lines]
        visited_island = [[False] * C for _ in range(R)]
        visited_bridge = [[False] * C for _ in range(R)]
        island_id_grid = [[0] * C for _ in range(R)]
        
        def is_valid(r, c):
            return 0 <= r < R and 0 <= c < C
        
        def bfs_island(r, c, id):
            queue = deque()
            queue.append((r, c))
            visited_island[r][c] = True
            island_id_grid[r][c] = id
            
            while queue:
                curr_r, curr_c = queue.popleft()
                
                for i in range(4):
                    nr = curr_r + dr[i]
                    nc = curr_c + dc[i]
                    
                    if is_valid(nr, nc) and not visited_island[nr][nc] and \
                       (grid[nr][nc] == '#' or grid[nr][nc] == 'X'):
                        visited_island[nr][nc] = True
                        island_id_grid[nr][nc] = id
                        queue.append((nr, nc))
        
        # 1. Identify Islands
        island_count = 0
        for r in range(R):
            for c in range(C):
                if not visited_island[r][c] and (grid[r][c] == '#' or grid[r][c] == 'X'):
                    island_count += 1
                    bfs_island(r, c, island_count)
        
        def bfs_find_islands_for_bridge(r, c):
            adjacent_island_ids = set()
            queue = deque()
            queue.append((r, c))
            visited_bridge[r][c] = True
            
            while queue:
                curr_r, curr_c = queue.popleft()
                
                for i in range(4):
                    nr = curr_r + dr[i]
                    nc = curr_c + dc[i]
                    
                    if is_valid(nr, nc):
                        if grid[nr][nc] == 'B' and not visited_bridge[nr][nc]:
                            visited_bridge[nr][nc] = True
                            queue.append((nr, nc))
                        elif grid[nr][nc] == 'X':
                            island_id = island_id_grid[nr][nc]
                            if island_id != 0:
                                adjacent_island_ids.add(island_id)
            return adjacent_island_ids
        
        # 2. Count Bridges & Connect Islands
        dsu = DSU(island_count)
        bridge_count = 0
        bus_count = island_count
        
        for r in range(R):
            for c in range(C):
                if grid[r][c] == 'B' and not visited_bridge[r][c]:
                    bridge_count += 1
                    connected_islands = bfs_find_islands_for_bridge(r, c)
                    
                    if len(connected_islands) == 2:
                        id1, id2 = list(connected_islands)
                        if dsu.find(id1) != dsu.find(id2):
                            dsu.unite(id1, id2)
                            bus_count -= 1
        
        print(f"Map {map_num}")
        print(f"islands: {island_count}")
        print(f"bridges: {bridge_count}")
        print(f"buses needed: {bus_count}")
        
        map_num += 1

if __name__ == "__main__":
    main()

File: 2013 ACM-ICPC North American Qualifier/island.java
import java.util.*;
import java.io.*;

public class island {
    static int R, C;
    static char[][] grid;
    static boolean[][] visitedIsland;
    static boolean[][] visitedBridge;
    static int[][] islandIdGrid;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    
    static class DSU {
        int[] parent;
        
        DSU(int n) {
            parent = new int[n + 1];
            for (int i = 0; i <= n; i++) {
                parent[i] = i;
            }
        }
        
        int find(int i) {
            if (parent[i] == i) return i;
            return parent[i] = find(parent[i]);
        }
        
        void unite(int i, int j) {
            int root_i = find(i);
            int root_j = find(j);
            if (root_i != root_j) {
                parent[root_i] = root_j;
            }
        }
    }
    
    static boolean isValid(int r, int c) {
        return r >= 0 && r < R && c >= 0 && c < C;
    }
    
    static void bfsIsland(int r, int c, int id) {
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{r, c});
        visitedIsland[r][c] = true;
        islandIdGrid[r][c] = id;
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int currR = curr[0], currC = curr[1];
            
            for (int i = 0; i < 4; i++) {
                int nr = currR + dr[i];
                int nc = currC + dc[i];
                
                if (isValid(nr, nc) && !visitedIsland[nr][nc] && 
                    (grid[nr][nc] == '#' || grid[nr][nc] == 'X')) {
                    visitedIsland[nr][nc] = true;
                    islandIdGrid[nr][nc] = id;
                    queue.offer(new int[]{nr, nc});
                }
            }
        }
    }
    
    static Set<Integer> bfsFindIslandsForBridge(int r, int c) {
        Set<Integer> adjacentIslandIds = new HashSet<>();
        Queue<int[]> queue = new LinkedList<>();
        
        queue.offer(new int[]{r, c});
        visitedBridge[r][c] = true;
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int currR = curr[0], currC = curr[1];
            
            for (int i = 0; i < 4; i++) {
                int nr = currR + dr[i];
                int nc = currC + dc[i];
                
                if (isValid(nr, nc)) {
                    if (grid[nr][nc] == 'B' && !visitedBridge[nr][nc]) {
                        visitedBridge[nr][nc] = true;
                        queue.offer(new int[]{nr, nc});
                    } else if (grid[nr][nc] == 'X') {
                        int islandId = islandIdGrid[nr][nc];
                        if (islandId != 0) {
                            adjacentIslandIds.add(islandId);
                        }
                    }
                }
            }
        }
        return adjacentIslandIds;
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line;
        int mapNum = 1;
        boolean firstMapOutput = true;
        
        while ((line = br.readLine()) != null) {
            while (line != null && line.isEmpty() && br.ready()) {
                line = br.readLine();
            }
            if (line == null) break;
            
            if (!firstMapOutput) {
                System.out.println();
            }
            firstMapOutput = false;
            
            List<String> gridLines = new ArrayList<>();
            gridLines.add(line);
            
            while ((line = br.readLine()) != null && !line.isEmpty()) {
                gridLines.add(line);
            }
            
            R = gridLines.size();
            C = gridLines.get(0).length();
            
            grid = new char[R][C];
            visitedIsland = new boolean[R][C];
            visitedBridge = new boolean[R][C];
            islandIdGrid = new int[R][C];
            
            for (int i = 0; i < R; i++) {
                grid[i] = gridLines.get(i).toCharArray();
            }
            
            // 1. Identify Islands
            int islandCount = 0;
            for (int r = 0; r < R; r++) {
                for (int c = 0; c < C; c++) {
                    if (!visitedIsland[r][c] && (grid[r][c] == '#' || grid[r][c] == 'X')) {
                        islandCount++;
                        bfsIsland(r, c, islandCount);
                    }
                }
            }
            
            // 2. Count Bridges & Connect Islands
            DSU dsu = new DSU(islandCount);
            int bridgeCount = 0;
            int busCount = islandCount;
            
            for (int r = 0; r < R; r++) {
                for (int c = 0; c < C; c++) {
                    if (grid[r][c] == 'B' && !visitedBridge[r][c]) {
                        bridgeCount++;
                        Set<Integer> connectedIslands = bfsFindIslandsForBridge(r, c);
                        
                        if (connectedIslands.size() == 2) {
                            Iterator<Integer> it = connectedIslands.iterator();
                            int id1 = it.next();
                            int id2 = it.next();
                            
                            if (dsu.find(id1) != dsu.find(id2)) {
                                dsu.unite(id1, id2);
                                busCount--;
                            }
                        }
                    }
                }
            }
            
            System.out.println("Map " + mapNum++);
            System.out.println("islands: " + islandCount);
            System.out.println("bridges: " + bridgeCount);
            System.out.println("buses needed: " + busCount);
        }
    }
}